Allow users to clear their profile and header picture

sternenseemann opened the pull request:
This Pull-Request is  #8646, but completely reworked.

ykzts reviewed (approved):
👍

ThibG commented:
It looks ok to me, however, we have recently merged #13192 which provides something similar for server-wide uploads. It would probably make sense to stick with that design.

Also note that one test fails because it clicks the first button of type `submit` to submit profile changes, but now, the first submit button deletes the header.

sternenseemann commented:
Thanks for the suggestion! I'll adjust it to that design, although I fear it might be a little fiddly to use on mobile? Also I'll try to get rid of the `type="submit"´

sternenseemann commented:
I updated the pull request to match the site upload deletion UI!

![Screenshot from 2020-03-31 18-57-11](https://user-images.githubusercontent.com/3154475/78055337-b30b6580-7383-11ea-9977-163f9c9da43e.png)

I tried to implement it similarly, but unfortunately I encountered the following problem: `@rails/ujs` only allows sending a request body when using `data-remote="true"` which makes the request asynchronous. Unfortunately here we need to send a `PUT` request with the body `account[subaction]=clear_avatar/header` in a synchronous fashion, so the settings page is reloaded and the user can see the changes. To fix this I implemented the missing feature myself, controlled by `data-sync-params` and `data-sync-method` in `utils/sync-ujs.js`.

I still feel like that this should be a part of `@rails/ujs`, so I [made a version](https://github.com/sternenseemann/rails/commit/ee79a0a41cc84bc5c896e06e3437ac82d12d60e9) of this for it as well and [suggested that change](https://discuss.rubyonrails.org/t/allow-data-params-for-ujs-links-that-dont-have-data-remote-true/73992). I hope this could be part of rails in the future, so mastodon could drop the own implementation of the feature.

ThibG commented:
The chunk of JS isn't that big, but wouldn't it be simpler to have separate routes for this instead of having new parameters?

Gargron commented on app/controllers/settings/profiles_controller.rb:
```diff
@@ -23,7 +23,7 @@ def update
 
   def account_params
-    params.require(:account).permit(:display_name, :note, :avatar, :header, :locked, :bot, :discoverable, fields_attributes: [:name, :value])
+    params.require(:account).permit(:display_name, :note, :avatar, :header, :locked, :bot, :discoverable, :subaction, fields_attributes: [:name, :value])
```
I think you kind of overcomplicated it. I think having separate routes for clearing avatar/header, and just POSTing to those routes through the `link_to` helper, and reloading the whole page would be fine.

sternenseemann committed:
Allow users to delete their header and avatar

This is achieved by sending a DELETE request to
/settings/profile/pictures/{avatar,header} via a link that is part of
the upload form's hint of the respective picture.

sternenseemann commented:
I've reworked the pull request completely and dropped the `subaction` parameter and the UI code completely. Instead of that I now send a simple `DELETE` request to `/settings/profile/pictures/avatar` or  `/settings/profile/pictures/header` similar to how it is done for site uploads.

At first I was thinking about adding a `/@username/avatar` etc. route which would show the avatar for every account and would allow to send a `DELETE` request for one's own account, but decided to keep everything in `/settings` for simplicity's sake. If you think that would be nicer, I could still rework it.

ThibG commented on app/helpers/settings_helper.rb:
```diff
@@ -105,4 +105,14 @@ def compact_account_link_to(account)
   end
+
+  def picture_hint(hint, picture)
+    link = link_to t('generic.delete'), settings_profile_picture_path(picture.name.to_s), data: { method: :delete }
+
+    if picture.original_filename.nil?
+      hint
+    else
+      safe_join([hint, link], '<br/>'.html_safe)
+    end
```
I'd move the `link` thing in the `else` statement.

sternenseemann replied:
Good point, fixed.

ClearlyClaire reviewed (changes requested):
LGTM overall, except for the part I commented inline

sternenseemann commented:
@ykzts @Gargron Is the change alright like this? Anything preventing a merge of this PR?
